File: iot-control/moony-android-app/app/src/main/java/com/moony/iotcontrol/DeviceAdapter.java
package com.moony.iotcontrol;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class DeviceAdapter extends RecyclerView.Adapter<DeviceAdapter.DeviceViewHolder> {

    private List<DeviceInfo> devices;

    public DeviceAdapter(List<DeviceInfo> devices) {
        this.devices = devices;
    }

    public void updateDevices(List<DeviceInfo> newDevices) {
        this.devices = newDevices;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public DeviceViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_device, parent, false);
        return new DeviceViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull DeviceViewHolder holder, int position) {
        DeviceInfo device = devices.get(position);
        holder.deviceName.setText(device.getDeviceName());
        holder.devicePlatform.setText(device.getPlatform());

        // Set appropriate icon based on platform
        String platform = device.getPlatform().toLowerCase();
        if (platform.contains("android")) {
            holder.deviceIcon.setText("📱");
        } else if (platform.contains("windows") || platform.contains("pc")) {
            holder.deviceIcon.setText("💻");
        } else if (platform.contains("mac") || platform.contains("ios")) {
            holder.deviceIcon.setText("🍎");
        } else if (platform.contains("linux")) {
            holder.deviceIcon.setText("🐧");
        } else {
            holder.deviceIcon.setText("📡");
        }
    }

    @Override
    public int getItemCount() {
        return devices != null ? devices.size() : 0;
    }

    static class DeviceViewHolder extends RecyclerView.ViewHolder {
        TextView deviceName;
        TextView devicePlatform;
        TextView deviceIcon;

        DeviceViewHolder(View itemView) {
            super(itemView);
            deviceName = itemView.findViewById(R.id.device_name);
            devicePlatform = itemView.findViewById(R.id.device_platform);
            deviceIcon = itemView.findViewById(R.id.device_icon);
        }
    }
}

File: iot-control/Moony-Android/app/src/main/java/com/moony/iot/adapter/DeviceAdapter.kt
package com.moony.iot.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.moony.iot.databinding.ItemDeviceBinding
import com.moony.iot.network.Device

class DeviceAdapter(private val onDeviceSelected: (Device) -> Unit) :
    ListAdapter<Device, DeviceAdapter.DeviceViewHolder>(DeviceDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DeviceViewHolder {
        val binding = ItemDeviceBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return DeviceViewHolder(binding, onDeviceSelected)
    }

    override fun onBindViewHolder(holder: DeviceViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    class DeviceViewHolder(
        private val binding: ItemDeviceBinding,
        private val onDeviceSelected: (Device) -> Unit
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(device: Device) {
            binding.apply {
                deviceNameText.text = device.deviceName
                deviceIdText.text = binding.root.context.getString(
                    com.moony.iot.R.string.device_id,
                    device.deviceId
                )
                devicePlatformText.text = binding.root.context.getString(
                    com.moony.iot.R.string.device_platform,
                    device.platform
                )

                if (!device.deviceType.isNullOrEmpty()) {
                    deviceTypeText.visibility = android.view.View.VISIBLE
                    deviceTypeText.text = binding.root.context.getString(
                        com.moony.iot.R.string.device_type,
                        device.deviceType
                    )
                } else {
                    deviceTypeText.visibility = android.view.View.GONE
                }

                selectButton.setOnClickListener {
                    onDeviceSelected(device)
                }

                root.setOnClickListener {
                    onDeviceSelected(device)
                }
            }
        }
    }

    private class DeviceDiffCallback : DiffUtil.ItemCallback<Device>() {
        override fun areItemsTheSame(oldItem: Device, newItem: Device): Boolean {
            return oldItem.deviceId == newItem.deviceId
        }

        override fun areContentsTheSame(oldItem: Device, newItem: Device): Boolean {
            return oldItem == newItem
        }
    }
}
